Brotli compression support

Theodlz opened the pull request:
Don't review yet, just a draft

stefanv commented on tools/fill_conf_values.py:
```diff
@@ -36,12 +36,35 @@ def hash_filter(string, htype):
 def fill_config_file_values(template_paths):
     log("Compiling configuration templates")
     env, cfg = load_env()
+    cfg["server"]["use_brotli"] = USE_BROTLI
```
The templates become quite confusing this way. Can we use `cfg["fill-config-feature"]` or similar?

Theodlz replied:
sounds good!

stefanv commented on tools/fill_conf_values.py:
```diff
@@ -36,12 +36,35 @@ def hash_filter(string, htype):
     log("Compiling configuration templates")
     env, cfg = load_env()
+    cfg["server"]["use_brotli"] = USE_BROTLI
+    cfg["server"]["is_macos"] = IS_MACOS
```
Probably not needed; just do all the logic for whether or not brotli should be loaded here.

Theodlz replied:
actually it is, because on mac once installed with homebrew brotli doesnt need to be dynamically loading in the nginx config. But if installed with linux it looks like you need to. So I do need to have both of these as separate arguments

Theodlz replied:
what I need to change is that I actually need to find the full path to the brotli module, which is trickier. But that way no need to figure out if its mac or not. I just need to find the path to where the nginx modules are stored and I'm done, I can always dynamically load brotli

Theodlz replied:
Yep, just got it to work on linux. I just need to verify that the dynamic import on macos is possible, and then I can have one single method for both.

Theodlz replied:
@stefanv I managed to install brotli both dynamically and statically in nginx while on linux. I think I've got it figured out. I'll add a piece of code which if statically installed, does not add the lines in the baselayer config to dynamically load brotli, and if installed dynamically, tries to figure out where the modules are so they can be loaded. If it can't figure out where the modules are but brotli is installed, just display a warning when running the app and start it without brotli

Theodlz commented:
@stefanv I need to make some changes to the GA (like go back to the real baselayer app template not my fork) before merging anything, but this works very well! The code snippet from the GA can be reused in the SkyPortal dockerfile to install nginx with brotli :). I will also add some documentation before merging, but re-requesting a review so you can see the new code.

stefanv commented on .github/workflows/test.yaml:
```diff
@@ -71,7 +74,25 @@ jobs:
+          sudo apt remove -y nginx nginx-common nginx-core
+
+          # install nginx from source so we can add the brotli module
+          git clone --recursive https://github.com/google/ngx_brotli.git
+          wget https://nginx.org/download/nginx-1.24.0.tar.gz
+          tar zxf nginx-1.24.0.tar.gz
+          cd ngx_brotli/deps/brotli
+          mkdir out && cd out
+          cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DCMAKE_C_FLAGS="-Ofast -m64 -march=native -mtune=native -flto -funroll-loops -ffunction-sections -fdata-sections -Wl,--gc-sections" -DCMAKE_CXX_FLAGS="-Ofast -m64 -march=native -mtune=native -flto -funroll-loops -ffunction-sections -fdata-sections -Wl,--gc-sections" -DCMAKE_INSTALL_PREFIX=./installed ..
+          cmake --build . --config Release --target brotlienc
+          cd ../../../..
+          export CURRENT_DIR=$(pwd)
+          cd nginx-1.24.0
+          ./configure --sbin-path=/usr/bin/nginx --conf-path=/usr/local/nginx/nginx.conf --pid-path=/usr/local/nginx/nginx.pid --with-http_ssl_module --with-stream --with-mail=dynamic --with-http_realip_module --with-compat --add-module=${CURRENT_DIR}/ngx_brotli
+          sudo make && sudo make install
+          sudo nginx
```
It's nice to be able to test this, but maybe a bit much for an optional feature.

Theodlz replied:
I agree, was just a good way for me to get it to run on github and see if it works as expected. Will remove it!

Maybe we can keep that in skyportal though, and add the test there

stefanv replied:
Or, if the simpler Debian instructions work, that'd be fine.

Theodlz replied:
Do you mean with libnginx-mod-http-brotli-filter? Well that was actually the issue for me, the lack of instruction. NGINX is supposed to be compile with the modules, so I didn't quite understand how installing this would work, nor what it did.

I can give it a go, but I found it more confusing than anything else.

stefanv replied:
I think you just need to install this package and then you can use `load_module modules/ngx_http_brotli_filter_module.so;`.

Theodlz replied:
if that works that's great! But then it means I can't detect whether or not brotli is installed dynamically by running `nginx -V`, which I suspect this package doesn't edit. Could work too, I'll give it a shot.

Theodlz replied:
@stefanv a simple apt install can't actually find `libnginx-mod-http-brotli-filter`. Did you install it differently?

stefanv commented on doc/setup.md:
````diff
@@ -37,7 +38,50 @@ See [below](#configuration) for more information on modifying the baselayer conf
+  - If you want to run NGINX as is (without brotli), you can simply install it with `sudo apt-get install nginx`.
+    If want to use [brotli compression](https://en.wikipedia.org/wiki/Brotli) with NGINX (to have better compression rates for the frontend), you can install NGINX with the `ngx_brotli` module with these commands:
+
+  ```
+  # install nginx from source so we can add the brotli module
+  git clone --recursive https://github.com/google/ngx_brotli.git
+  wget https://nginx.org/download/nginx-1.24.0.tar.gz
+  tar zxf nginx-1.24.0.tar.gz
+  cd ngx_brotli/deps/brotli
+  mkdir out && cd out
+  cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DCMAKE_C_FLAGS="-Ofast -m64 -march=native -mtune=native -flto -funroll-loops -ffunction-sections -fdata-sections -Wl,--gc-sections" -DCMAKE_CXX_FLAGS="-Ofast -m64 -march=native -mtune=native -flto -funroll-loops -ffunction-sections -fdata-sections -Wl,--gc-sections" -DCMAKE_INSTALL_PREFIX=./installed ..
+  cmake --build . --config Release --target brotlienc
+  cd ../../../..
+  export CURRENT_DIR=$(pwd)
+  cd nginx-1.24.0
+  ./configure --sbin-path=/usr/sbin/nginx --conf-path=/usr/local/nginx/nginx.conf --pid-path=/usr/local/nginx/nginx.pid --with-http_ssl_module --with-stream --with-mail=dynamic --with-http_realip_module --with-compat --add-module=${CURRENT_DIR}/ngx_brotli
+  sudo make && sudo make install
+  ```
````
What's wrong with https://packages.debian.org/sid/libnginx-mod-http-brotli-filter?

stefanv commented on tools/fill_conf_values.py:
```diff
@@ -36,12 +37,76 @@ def hash_filter(string, htype):
+
+    Returns
+    -------
+    installed : bool
+        True if the nginx brotli module is installed, False otherwise
+    dynamic : bool
+        True if the module is dynamically loaded, False otherwise
+    modules_dir : str
+        The directory where the nginx modules are located if dynamic loading is used
```
Somehow, a NamedTuple feels like it might work well here, but that's total nitpicking and not even clearly an improvement :)

Theodlz replied:
hahaha no of course. I'll try that if I have the time.

stefanv commented on tools/fill_conf_values.py:
```diff
@@ -36,12 +37,76 @@ def hash_filter(string, htype):
+            # we try to figure out where the modules are, there are 2 possibilities
+            # 1. the --modules-path directive is used
+            # 2. the default directory is used, which is where the configuration file is located so we look for it
+            if "--modules-path" in output:
```
nginx has a search path, so will

```
load_module modules/ngx_http_brotli_filter_module.so;
load_module ...
```

not suffice?

Theodlz replied:
Well I actually tried that on my linux machine, and it didn't work! Not sure why... but it did look into the directory where the config is and didn't look at any other path.

stefanv replied:
Do you have a minimal config you use to test with?

Theodlz replied:
What do you mean? I basically run SkyPortal with that modified version of baselayer. I ran it on an ubuntu machine with 8GB of ram, that had nginx but that I replaced with these instructions.

stefanv reviewed (commented):
Looks good! A few small questions.

Theodlz commented:
@stefanv updated the GA and the docs to install from the ppa, and updated the python logic to find the nginx brotli module. This is solid, works well on both dynamic and static installations (and on both windows and mac).

Let me know what you think. If we want this tested, we will have to update the baselayer template app with a test (like the one I added to the fork pinned here for testing). So if you are happy with these changes, let me know and I'll pin the normal template app again and we can merge :)

stefanv commented:
I like it! Let's also merge the test on BL.

Theodlz commented:
Great! I went back to git cloning the real template app in the GA. I'll open the PR on the template app ASAP.
